swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift

```diff
@@ -38,89 +38,87 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
     super.startTunnel(options: options, completionHandler: completionHandler)
     Log.log("\(#function)")
 
-    Task {
-      // If we don't have a token, we can't continue.
-      guard let token = loadAndSaveToken(from: options)
-      else {
-        completionHandler(PacketTunnelProviderError.tokenNotFoundInKeychain)
+    // If we don't have a token, we can't continue.
+    guard let token = loadAndSaveToken(from: options)
+    else {
+      completionHandler(PacketTunnelProviderError.tokenNotFoundInKeychain)
 
-        return
-      }
+      return
```

Comment:
This function repeats the pattern of "call completion handler and early return" quite often.

I think this could be simplified if we had a private function that simply throws exceptions and this function here does a single `do try catch` and then calls the completion handler.

Reply:
That's what we did before but then we would fail to connect if for example we couldn't save the token to the Keychain, which is the purpose of the stacked PR.

But yes in a block where we want to bail on _any_ error I think that would be simpler.

I realized in that PR that do...try is syntactically heavy if you want to act on some error and continue. The do block doesn't return its contents and encapsulates variable scope so you can't simply do:

```
let foo: String = do { try read() }
catch { log() }
finally { "can't read" }
```

like you can in some other languages :-/.

Reply:
> That's what we did before but then we would fail to connect if for example we couldn't save the token to the Keychain, which is the purpose of the stacked PR.

This should be solvable with another, nested `do try` like you do in the current PR. The outer control flow shouldn't matter for that.

Reply:
Ah yes, I see what you mean. So you're saying use `throw` instead of `guard...else`.

Reply:
> So you're saying use `throw` instead of `guard...else`.

Yes. `throw` inside another helper function which gets called from `startTunnel` with a `do try` and passes the error to the completion handler. That will also mean that we only have a single place to call `Log.error` though.